src/exojax/spec/plg.py

```diff
@@ -0,0 +1,427 @@
+    Nline=len(elower)
+    Nunf=np.sum(~frozen_mask)
+    Npl=len(qlogsij0[nonzeropl_mask])
+    if verbose:
+        print("# of original lines:",Nline)
+        print("# of unfrozen lines:",Nunf)
+        print("# of frozen lines:",np.sum(frozen_mask))
+        print("# of pseudo lines:",Npl)
+        print("# of total lines:",(Npl+Nunf))
+        print("# compression:",(Npl+Nunf)/Nline)
+
+    if reshape==True:
+        qlogsij0=qlogsij0.reshape(Na,Nnugrid,Nelower)
+        qcnu=qcnu.reshape(Na,Nnugrid,Nelower)
+        num_unique=num_unique.reshape(Na,Nnugrid,Nelower)
+        
+    return qlogsij0,qcnu,num_unique,elower_grid,frozen_mask,nonzeropl_mask
+
+
+def get_qlogsij0_addcon(indexa,Na,cnu,indexnu,Nnugrid,mdb,expme,expme_grid, Ncrit=0, Tuplim=1000., threshold_persist_freezing=10000):
+    """gather (freeze) lines w/ additional indexing
+
+    Args:
+       indexa: the indexing of the additional condition
+       Na: the number of the additional condition grid
+       cnu: contribution of wavenumber for LSD
+       indexnu: index of wavenumber
+       Nnugrid: number of nu grid
+       mdb: molecular database (instance made by the MdbExomol/MdbHit class in moldb.py)
+       expme: exp(- hcperk*(elower/Tuplimc - elower/Tref) - preov)
+       expme_grid: grid of expme
+       Ncrit: frozen line number per bin
+       Tuplim: rough guess on the typical atmospheric temperature
+       threshold_persist_freezing: How weak lines compared to the deepest one will you persist freezing (not retaken)?
+
+    Returns:
+        qlogsij0: log line strength
+        qcnu: cnu for pseudo line (contribution of wavenumber for LSD)
+        num_unique: number of the lines in each bin of grid
+        frozen_mask: mask for frozen lines into pseudo lines
+
+    """
+    m=len(expme_grid)
+    n=m*Nnugrid
+    Ng=n*Na
+    
+    cont,index=npgetix(expme,expme_grid) #elower contribution and elower index of lines
+    eindex=index+m*indexnu+n*indexa #extended index elower,nu,a
+    
+    #frozen criterion
+    num_unique=np.bincount(eindex,minlength=Ng) # number of the lines in a bin
+    
+    lmask=(num_unique>=Ncrit)
+    erange=range(0,Ng)
+    frozen_eindex=np.array(erange)[lmask]
+    frozen_mask=np.isin(eindex,frozen_eindex)
+    
+    #Retake relatively strong lines.
+    SijTgue_frozen = SijT(Tuplim+500., \
+                    mdb.logsij0[frozen_mask], mdb.nu_lines[frozen_mask], mdb.elower[frozen_mask], \
+                    qT=mdb.qr_interp(Tuplim+500.))
+    persist_freezing = SijTgue_frozen < max(SijTgue_frozen) / threshold_persist_freezing
+    index_persist_freezing = np.where(frozen_mask)[0][persist_freezing]
+    frozen_mask = np.isin(np.arange(len(frozen_mask)), index_persist_freezing)
+        
+    #qlogsij0
+    Sij=np.exp(mdb.logsij0)
+    qlogsij0=np.bincount(eindex,weights=Sij*(1.0-cont)*frozen_mask,minlength=Ng)
+    qlogsij0=qlogsij0+np.bincount(eindex+1,weights=Sij*cont*frozen_mask,minlength=Ng)
+    with warnings.catch_warnings():
```

Comment:
Just comment: cool. I should have used `warnings` package in the code I wrote. I will do that in the forthcoming refactoring